meshparty/trimesh_io.py

```diff
@@ -861,25 +860,23 @@ def filter_unmasked_boolean(self, unmasked_boolean):
         '''
         return unmasked_boolean[self.node_mask]
 
-    def filter_unmasked_indices(self, unmasked_shape, include_outside=False):
-        '''
-        For an array of indices in the original mesh, returns the indices filtered and remapped
-        for the masked mesh. Nans out rows not in the mask. Preserves the order of unmasked_indices.
-        '''
-        filtered_shape = utils.filter_shapes(self.indices_unmasked, unmasked_shape)[0]
+    def filter_unmasked_indices(self, unmasked_shape, mask=None, include_outside=False):
+        if mask is None:
+            mask = self.node_mask
+        new_index = np.zeros(mask.shape)-1
+        new_index[mask] = np.arange(np.sum(mask))
+        new_shape = new_index[unmasked_shape.ravel()].reshape(unmasked_shape.shape).astype(int)
+        
+        if len(new_shape.shape) > 1:
+            keep_rows = np.all(new_shape>=0, axis=1)
+        else:
+            keep_rows = new_shape>=0
+
         if include_outside:
-            long_shape = unmasked_shape.ravel()
-            within_mask = self.node_mask[long_shape].reshape(unmasked_shape.shape)
-            filtered_shape_all = np.full(unmasked_shape.shape, np.nan)
-            if within_mask.ndim==1:
-                filtered_shape_all[within_mask==True] = filtered_shape.squeeze()
-            else:
-                filtered_shape_all[np.all(within_mask==True, axis=1)] = filtered_shape
-            filtered_shape = filtered_shape_all.astype(int).squeeze()
+            new_shape[~keep_rows] = np.nan
```

Comment:
we need a test here.

Reply:
Better to omit it entirely, I think. It's easy to replicate the overall goal (to know which indices are not inside the mask) in other ways and I have never used it.